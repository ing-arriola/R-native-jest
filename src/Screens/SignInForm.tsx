import React from 'react';
import {StyleSheet, Text, KeyboardAvoidingView} from 'react-native';
import ErrorText from '../Components/ErrorText';
import {useLoginFormState} from '../Hooks/useLoginFormState';
import Input from '../Components/Input';
import Button from '../Components/Button';
import {StackScreenProps} from '@react-navigation/stack';
import {RootStackParamList} from '../../App';

interface Props extends StackScreenProps<RootStackParamList, 'signInForm'> {}

const SignInForm = ({navigation}: Props) => {
  const {username, password, submit} = useLoginFormState();
  let usernameErrorMsg = '';
  let passwordErrorMsg = '';

  if (submit.value && !username.valid) {
    usernameErrorMsg = 'Invalid username.';
  }

  if (submit.value && !password.valid) {
    passwordErrorMsg = 'Invalid password.';
  }

  return (
    <KeyboardAvoidingView style={styles.container} behavior="position">
      <Text style={styles.headerText}>Login</Text>
      <Input
        label="Username"
        placeholder="example"
        onChangeText={username.set}
        error={usernameErrorMsg}
        testID="SignIn.usernameInput"
        secureTextEntry={false}
      />
      <Input
        label="Password"
        placeholder="***"
        secureTextEntry
        onChangeText={password.set}
        error={passwordErrorMsg}
        testID="SignIn.passwordInput"
      />
      <ErrorText messages={[usernameErrorMsg, passwordErrorMsg]} />
      <Button
        testID="SignIn.Button"
        text="Login"
        onPress={() => submit.set()}
      />
    </KeyboardAvoidingView>
  );
};

export default SignInForm;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    justifyContent: 'center',
    paddingHorizontal: 40,
  },
  headerText: {
    color: '#353031',
    fontWeight: 'bold',
    fontSize: 34,
    marginBottom: 10,
  },
});
