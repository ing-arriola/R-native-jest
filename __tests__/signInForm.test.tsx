import React from 'react';
import {fireEvent, render, waitFor} from '@testing-library/react-native';
import SignInForm, {Props} from '../src/Screens/SignInForm';

const mockedNavigate = jest.fn();

describe('Test for signInForm', () => {
  let props: Props;

  beforeEach(() => {
    props = {
      navigation: {navigate: mockedNavigate} as any,
    };
  });

  it('should render', () => {
    const {toJSON} = render(<SignInForm {...props} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should have 2 login texts', () => {
    const {getAllByText} = render(<SignInForm {...props} />);
    expect(getAllByText('Login').length).toBe(2);
  });

  it('should show error messages', () => {
    const {getByTestId, getByText} = render(<SignInForm {...props} />);
    fireEvent.press(getByTestId('SignIn.Button'));
    getByText('Invalid username.');
    getByText('Invalid password.');
  });

  it('should show invalid username message', () => {
    const {getByTestId, getByText, queryAllByText} = render(
      <SignInForm {...props} />,
    );
    fireEvent.changeText(getByTestId('SignIn.usernameInput'), 'wrong');
    fireEvent.changeText(getByTestId('SignIn.passwordInput'), 'asd');
    fireEvent.press(getByTestId('SignIn.Button'));
    getByText('Invalid username.');
    expect(queryAllByText('Invalid password.').length).toBe(0);
  });

  it('should show invalid password message', () => {
    const {getByTestId, getByText, queryAllByText} = render(
      <SignInForm {...props} />,
    );
    fireEvent.changeText(getByTestId('SignIn.usernameInput'), 'asd');
    fireEvent.changeText(getByTestId('SignIn.passwordInput'), 'wrong');
    fireEvent.press(getByTestId('SignIn.Button'));
    getByText('Invalid password.');
    expect(queryAllByText('Invalid username.').length).toBe(0);
  });

  it('should navigate Home', async () => {
    const {getByTestId} = render(<SignInForm {...props} />);

    fireEvent.changeText(getByTestId('SignIn.usernameInput'), 'asd');
    fireEvent.changeText(getByTestId('SignIn.passwordInput'), 'asd');
    fireEvent.press(getByTestId('SignIn.Button'));

    await waitFor(() => {
      expect(mockedNavigate).toBeCalledWith('Home');
    });
  });
});
