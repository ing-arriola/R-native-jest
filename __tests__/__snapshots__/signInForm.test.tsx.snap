// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for signInForm should render 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 40,
    }
  }
>
  <View
    style={
      Object {
        "bottom": 0,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#353031",
          "fontSize": 34,
          "fontWeight": "bold",
          "marginBottom": 10,
        }
      }
    >
      Login
    </Text>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#f4f6f8",
            "borderColor": "#f4f6f8",
            "borderRadius": 10,
            "borderWidth": 1,
            "marginVertical": 5,
            "paddingHorizontal": 15,
            "paddingVertical": 10,
          },
          null,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#b4b6b8",
            "fontSize": 10,
          }
        }
      >
        Username
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <TextInput
          autoCapitalize="none"
          onChangeText={[Function]}
          secureTextEntry={false}
          style={
            Object {
              "color": "#353031",
              "flex": 1,
              "fontSize": 14,
              "fontWeight": "bold",
              "marginRight": 10,
              "marginTop": 3,
            }
          }
          testID="SignIn.usernameInput"
        />
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#f4f6f8",
            "borderColor": "#f4f6f8",
            "borderRadius": 10,
            "borderWidth": 1,
            "marginVertical": 5,
            "paddingHorizontal": 15,
            "paddingVertical": 10,
          },
          null,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#b4b6b8",
            "fontSize": 10,
          }
        }
      >
        Password
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <TextInput
          autoCapitalize="none"
          onChangeText={[Function]}
          secureTextEntry={true}
          style={
            Object {
              "color": "#353031",
              "flex": 1,
              "fontSize": 14,
              "fontWeight": "bold",
              "marginRight": 10,
              "marginTop": 3,
            }
          }
          testID="SignIn.passwordInput"
        />
      </View>
    </View>
    <View
      style={
        Object {
          "paddingVertical": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#cc0011",
            "fontSize": 14,
          }
        }
      >
        
      </Text>
      <Text
        style={
          Object {
            "color": "#cc0011",
            "fontSize": 14,
          }
        }
      >
        
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#9374b7",
          "borderRadius": 10,
          "justifyContent": "center",
          "marginTop": 10,
          "opacity": 1,
          "paddingVertical": 16,
        }
      }
      testID="SignIn.Button"
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        Login
      </Text>
    </View>
  </View>
</View>
`;
